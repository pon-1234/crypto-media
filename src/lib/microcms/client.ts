/**
 * microCMS クライアント設定
 * @doc DEVELOPMENT_GUIDE.md#microCMS
 * @issue #2 - microCMSクライアントと型定義の実装
 */
import { createClient } from 'microcms-js-sdk'

if (!process.env.MICROCMS_SERVICE_DOMAIN) {
  throw new Error('MICROCMS_SERVICE_DOMAIN is required')
}

if (!process.env.MICROCMS_API_KEY) {
  throw new Error('MICROCMS_API_KEY is required')
}

/**
 * microCMS クライアントインスタンス
 * @see https://github.com/microcmsio/microcms-js-sdk
 */
export const client = createClient({
  serviceDomain: process.env.MICROCMS_SERVICE_DOMAIN,
  apiKey: process.env.MICROCMS_API_KEY,
})

/**
 * プレビューモードでの記事取得について
 * 
 * プレビューモードでは通常のクライアントを使用し、
 * APIメソッド呼び出し時にdraftKeyをクエリパラメータとして渡します。
 * 
 * 例:
 * ```typescript
 * const preview = await client.get({
 *   endpoint: 'media_articles',
 *   contentId: 'xxx',
 *   queries: { draftKey: 'yyy' }
 * })
 * ```
 */

/**
 * API呼び出しの共通オプション
 */
export const defaultQueries = {
  limit: 100,
  orders: '-publishedAt', // 最新順
} as const

/**
 * リスト取得時の最大件数
 */
export const MAX_LIMIT = 100

/**
 * 画像最適化用のパラメータを生成
 * @param url - microCMSの画像URL
 * @param options - 画像変換オプション
 */
export const getOptimizedImageUrl = (
  url: string,
  options: {
    width?: number
    height?: number
    format?: 'webp' | 'jpg' | 'png'
    quality?: number
  } = {}
): string => {
  const params = new URLSearchParams()
  
  if (options.width) params.append('w', options.width.toString())
  if (options.height) params.append('h', options.height.toString())
  if (options.format) params.append('fm', options.format)
  if (options.quality) params.append('q', options.quality.toString())
  
  return `${url}?${params.toString()}`
}