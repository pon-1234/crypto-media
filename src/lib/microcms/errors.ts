/**
 * @doc microCMS error handling utilities
 * @issue Gemini review recommendation - Better error handling
 */

/**
 * microCMS API error class
 */
export class MicroCMSApiError extends Error {
  constructor(
    message: string,
    public status: number,
    public code?: string
  ) {
    super(message)
    this.name = 'MicroCMSApiError'
  }
}

/**
 * Check if error is a microCMS 404 error
 */
export function isMicroCMS404Error(error: unknown): boolean {
  if (error instanceof MicroCMSApiError) {
    return error.status === 404
  }
  
  // Fallback for microCMS SDK errors
  if (error instanceof Error) {
    // Check for common 404 patterns in error messages
    const message = error.message.toLowerCase()
    return (
      message.includes('404') ||
      message.includes('not found') ||
      message.includes('content not found')
    )
  }
  
  return false
}

/**
 * Check if error is a microCMS API error
 */
export function isMicroCMSError(error: unknown): error is MicroCMSApiError {
  return error instanceof MicroCMSApiError
}

/**
 * Parse microCMS error response
 */
export function parseMicroCMSError(error: unknown): MicroCMSApiError {
  // If it's already a MicroCMSApiError, return it as-is
  if (error instanceof MicroCMSApiError) {
    return error
  }
  
  // Handle Axios-like errors with response property
  if (
    error &&
    typeof error === 'object' &&
    'response' in error &&
    error.response &&
    typeof error.response === 'object' &&
    'status' in error.response
  ) {
    const response = error.response as { status: number; data?: { message?: string } }
    return new MicroCMSApiError(
      response.data?.message || 'API Error',
      response.status
    )
  }

  // Handle standard Error objects
  if (error instanceof Error) {
    // Try to extract status from error message
    const statusMatch = error.message.match(/(\d{3})/)
    const status = statusMatch ? parseInt(statusMatch[1], 10) : 500
    
    return new MicroCMSApiError(error.message, status)
  }

  // Fallback for unknown errors
  return new MicroCMSApiError('Unknown error occurred', 500)
}