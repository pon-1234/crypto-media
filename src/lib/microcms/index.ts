/**
 * microCMS ライブラリのエントリーポイント
 * @doc DEVELOPMENT_GUIDE.md#microCMS
 * @issue #2 - microCMSクライアントと型定義の実装
 */

// クライアント
export { client, defaultQueries, MAX_LIMIT, getOptimizedImageUrl } from './client'

// スキーマとバリデーション
export {
  // スキーマ
  corporateNewsSchema,
  expertSchema,
  categorySchema,
  tagSchema,
  featureSchema,
  mediaArticleSchema,
  siteSettingsSchema,
  researchSurveySchema,
  pageCorporateSchema,
  
  // リストレスポンススキーマ
  corporateNewsListSchema,
  expertListSchema,
  categoryListSchema,
  tagListSchema,
  featureListSchema,
  mediaArticleListSchema,
  researchSurveyListSchema,
  pageCorporateListSchema,
  
  // ヘルパー関数
  createListResponseSchema,
  
  // 型
  type CorporateNews,
  type Expert,
  type Category,
  type Tag,
  type Feature,
  type MediaArticle,
  type SiteSettings,
  type ResearchSurvey,
  type PageCorporate,
} from '@/lib/schema'

// 追加の型エクスポート
export type {
  MicroCMSBase,
  MicroCMSListResponse,
  MicroCMSImage,
  ExpertRole,
  ArticleType,
  MembershipLevel,
} from '@/lib/schema'

// APIメソッド
export {
  getCorporateNewsList,
  getCorporateNewsDetail,
  getAllCorporateNewsIds,
} from './corporate-news'