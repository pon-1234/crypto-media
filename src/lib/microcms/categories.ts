import { client } from './client'
import type { Category } from '@/lib/schema/category.schema'
import type {
  MicroCMSQueries,
  MicroCMSListResponse,
  MicroCMSContentId,
} from 'microcms-js-sdk'

/**
 * Get all categories
 *
 * @doc Fetches all available categories from microCMS
 * @related src/lib/schema/category.schema.ts - Category type definition
 * @param queries - Optional microCMS query parameters
 * @returns Promise containing list of categories
 */
export const getCategories = async (
  queries?: MicroCMSQueries
): Promise<MicroCMSListResponse<Category>> => {
  return await client.getList<Category>({
    endpoint: 'categories',
    queries,
  })
}

/**
 * Get a single category by slug
 *
 * @doc Fetches a specific category by its slug identifier
 * @related src/lib/schema/category.schema.ts - Category type definition
 * @param slug - The category slug to fetch
 * @returns Promise containing the category data or null if not found
 */
export const getCategoryBySlug = async (
  slug: string
): Promise<Category | null> => {
  try {
    const response = await client.getList<Category>({
      endpoint: 'categories',
      queries: {
        filters: `slug[equals]${slug}`,
        limit: 1,
      },
    })
    return response.contents[0] || null
  } catch (error) {
    console.error('Error fetching category by slug:', error)
    return null
  }
}

/**
 * Get a single category by ID
 *
 * @doc Fetches a specific category by its microCMS ID
 * @related src/lib/schema/category.schema.ts - Category type definition
 * @param id - The microCMS content ID
 * @param queries - Optional microCMS query parameters
 * @returns Promise containing the category data
 */
export const getCategoryById = async (
  id: string,
  queries?: MicroCMSQueries
): Promise<Category & MicroCMSContentId> => {
  return await client.get<Category>({
    endpoint: 'categories',
    contentId: id,
    queries,
  })
}