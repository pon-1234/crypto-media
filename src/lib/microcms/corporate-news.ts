import {
  corporateNewsListSchema,
  corporateNewsSchema,
  type CorporateNews,
} from '@/lib/schema'
import type { MicroCMSQueries } from 'microcms-js-sdk'
import { client } from './client'
import { handleError } from '@/lib/utils/handleError'

/**
 * コーポレートお知らせ一覧を取得する
 *
 * @issue #4 - コーポレートお知らせ一覧・詳細ページの実装
 */
export async function getCorporateNewsList(queries?: MicroCMSQueries): Promise<{
  contents: CorporateNews[]
  totalCount: number
  offset: number
  limit: number
}> {
  try {
    const response = await client.get({
      endpoint: 'corporate_news',
      queries,
    })

    return corporateNewsListSchema.parse(response)
  } catch (error) {
    handleError(error, 'Failed to fetch corporate news list')
    throw error
  }
}

/**
 * コーポレートお知らせの詳細を取得する
 *
 * @issue #4 - コーポレートお知らせ一覧・詳細ページの実装
 */
export async function getCorporateNewsDetail(
  contentId: string,
  queries?: MicroCMSQueries
): Promise<CorporateNews> {
  try {
    const response = await client.get({
      endpoint: 'corporate_news',
      contentId,
      queries,
    })

    return corporateNewsSchema.parse(response)
  } catch (error) {
    handleError(
      error,
      `Failed to fetch corporate news detail for id: ${contentId}`
    )
    throw error
  }
}

/**
 * すべてのコーポレートお知らせのIDを取得する
 * generateStaticParams用
 *
 * @issue #4 - コーポレートお知らせ一覧・詳細ページの実装
 */
export async function getAllCorporateNewsIds(): Promise<string[]> {
  try {
    const response = await client.get({
      endpoint: 'corporate_news',
      queries: {
        fields: 'id',
        limit: 100, // 必要に応じて調整
      },
    })

    // fields: 'id'を指定した場合は、idのみが返るためスキーマ検証をスキップ
    return response.contents.map((item: { id: string }) => item.id)
  } catch (error) {
    handleError(error, 'Failed to fetch corporate news IDs')
    throw error
  }
}

/**
 * カテゴリ別にコーポレートお知らせ一覧を取得する
 * TODO: microCMSにcategoryフィールドを追加後に実装を有効化
 *
 * @issue #35 - コーポレートサイトの未実装ページ作成
 */
export async function getCorporateNewsListByCategory(
  category: string,
  queries?: MicroCMSQueries
): Promise<{
  contents: CorporateNews[]
  totalCount: number
  offset: number
  limit: number
}> {
  try {
    // TODO: categoryフィールド追加後に以下のコメントを解除
    // const response = await client.get({
    //   endpoint: 'corporate_news',
    //   queries: {
    //     ...queries,
    //     filters: `category[equals]${category}`,
    //   },
    // })

    // 仮実装: 通常の一覧を返す
    const response = await client.get({
      endpoint: 'corporate_news',
      queries,
    })

    return corporateNewsListSchema.parse(response)
  } catch (error) {
    handleError(
      error,
      `Failed to fetch corporate news list for category: ${category}`
    )
    throw error
  }
}
