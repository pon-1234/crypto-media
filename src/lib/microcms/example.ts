/**
 * microCMS APIの使用例
 * @doc DEVELOPMENT_GUIDE.md#microCMS
 * @issue #2 - microCMSクライアントと型定義の実装
 */
import { client, defaultQueries, getOptimizedImageUrl } from './client'
import {
  mediaArticleSchema,
  mediaArticleListSchema,
  siteSettingsSchema,
  type MediaArticle,
  type SiteSettings,
} from './index'

/**
 * 記事一覧を取得する例
 */
export async function getArticles(limit = 10): Promise<MediaArticle[]> {
  try {
    const response = await client.get({
      endpoint: 'media_articles',
      queries: {
        ...defaultQueries,
        limit,
      },
    })

    // zodでバリデーション
    const validated = mediaArticleListSchema.parse(response)
    return validated.contents
  } catch (error) {
    console.error('Failed to fetch articles:', error)
    throw error
  }
}

/**
 * 特定の記事を取得する例（改善版）
 * slugで記事を検索し、見つからない場合はnullを返す
 */
export async function getArticleBySlug(
  slug: string
): Promise<MediaArticle | null> {
  try {
    const response = await client.get({
      endpoint: 'media_articles',
      queries: {
        filters: `slug[equals]${slug}`,
        limit: 1,
      },
    })

    const validated = mediaArticleListSchema.parse(response)

    // 記事が見つからない場合
    if (validated.totalCount === 0) {
      return null
    }

    // 念のため、totalCountが1であることを確認
    if (validated.totalCount !== 1) {
      console.warn(
        `Expected 1 article with slug "${slug}", but found ${validated.totalCount}`
      )
    }

    return validated.contents[0]
  } catch (error) {
    // microCMSのエラーレスポンスを適切に処理
    if (error instanceof Error && 'response' in error) {
      const httpError = error as Error & { response?: { status: number } }
      if (httpError.response?.status === 404) {
        return null
      }
    }

    console.error('Failed to fetch article:', error)
    throw error
  }
}

/**
 * プレビュー用に記事を取得する例
 */
export async function getArticlePreview(
  contentId: string,
  draftKey: string
): Promise<MediaArticle> {
  try {
    const response = await client.get({
      endpoint: 'media_articles',
      contentId,
      queries: {
        draftKey,
      },
    })

    // 単一記事のバリデーション
    const validated = mediaArticleSchema.parse(response)
    return validated
  } catch (error) {
    console.error('Failed to fetch preview:', error)
    throw error
  }
}

/**
 * 有料会員向け記事のみを取得する例
 */
export async function getPaidArticles(): Promise<MediaArticle[]> {
  try {
    const response = await client.get({
      endpoint: 'media_articles',
      queries: {
        ...defaultQueries,
        filters: 'membershipLevel[equals]paid',
      },
    })

    const validated = mediaArticleListSchema.parse(response)
    return validated.contents
  } catch (error) {
    console.error('Failed to fetch paid articles:', error)
    throw error
  }
}

/**
 * 画像URLを最適化する例
 */
export function getOptimizedArticleImage(article: MediaArticle) {
  if (!article.heroImage) return null

  return {
    // デスクトップ用
    desktop: getOptimizedImageUrl(article.heroImage.url, {
      width: 1200,
      format: 'webp',
      quality: 85,
    }),
    // モバイル用
    mobile: getOptimizedImageUrl(article.heroImage.url, {
      width: 600,
      format: 'webp',
      quality: 85,
    }),
    // OGP用
    ogp: getOptimizedImageUrl(article.heroImage.url, {
      width: 1200,
      height: 630,
      format: 'jpg',
      quality: 90,
    }),
  }
}

/**
 * サイト設定を取得する例（シングルトンAPI）
 */
export async function getSiteSettings(): Promise<SiteSettings> {
  try {
    const response = await client.get({
      endpoint: 'site_settings',
    })

    // シングルトンAPIはオブジェクトを直接返す
    const validated = siteSettingsSchema.parse(response)
    return validated
  } catch (error) {
    console.error('Failed to fetch site settings:', error)
    throw error
  }
}
