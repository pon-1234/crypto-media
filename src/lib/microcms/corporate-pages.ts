/**
 * @doc Corporate pages API client methods
 * @related DEVELOPMENT_GUIDE.md#pages_corporate
 * @issue https://github.com/pon-1234/crypto-media/issues/25
 */
import { client, defaultQueries } from './client'
import {
  corporatePageSchema,
  corporatePageListSchema,
  type CorporatePage,
  type CorporatePageList,
} from '@/lib/schema/corporate-page.schema'
import { isMicroCMS404Error, parseMicroCMSError } from './errors'

/**
 * Corporate page API query parameters
 */
type CorporatePageQueries = {
  draftKey?: string
  limit?: number
  offset?: number
}

/**
 * Get all corporate pages
 */
export async function getCorporatePages(
  queries?: CorporatePageQueries
): Promise<CorporatePageList> {
  const response = await client.get({
    endpoint: 'pages_corporate',
    queries: {
      ...defaultQueries,
      ...queries,
    },
  })

  return corporatePageListSchema.parse(response)
}

/**
 * Get a single corporate page by slug
 */
export async function getCorporatePageBySlug(
  slug: string,
  queries?: Pick<CorporatePageQueries, 'draftKey'>
): Promise<CorporatePage | null> {
  try {
    const response = await client.get({
      endpoint: 'pages_corporate',
      queries: {
        filters: `slug[equals]${slug}`,
        limit: 1,
        ...queries,
      },
    })

    const parsed = corporatePageListSchema.parse(response)
    return parsed.contents[0] || null
  } catch (error) {
    const microCMSError = parseMicroCMSError(error)
    if (isMicroCMS404Error(microCMSError)) {
      return null
    }
    throw microCMSError
  }
}

/**
 * Get a single corporate page by ID (for preview mode)
 */
export async function getCorporatePageById(
  id: string,
  queries?: Pick<CorporatePageQueries, 'draftKey'>
): Promise<CorporatePage | null> {
  try {
    const response = await client.get({
      endpoint: 'pages_corporate',
      contentId: id,
      queries,
    })

    return corporatePageSchema.parse(response)
  } catch (error) {
    const microCMSError = parseMicroCMSError(error)
    if (isMicroCMS404Error(microCMSError)) {
      return null
    }
    throw microCMSError
  }
}

/**
 * Check if a corporate page exists by slug
 */
export async function corporatePageExists(slug: string): Promise<boolean> {
  const page = await getCorporatePageBySlug(slug)
  return page !== null
}
