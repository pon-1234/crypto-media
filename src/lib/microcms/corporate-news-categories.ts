/**
 * コーポレートニュースカテゴリ用のAPIクライアント
 * @doc コーポレートニュースのカテゴリ管理機能
 * @related corporate-news.ts
 * @issue #41 - コーポレートニュースのカテゴリ機能
 */
import { client } from './client'
import { corporateNewsCategoryListSchema } from '@/lib/schema'
import { getAllContents } from './utils'
import type { CorporateNewsCategory } from '@/lib/schema'

/**
 * コーポレートニュースカテゴリをスラッグで取得
 * @doc 指定されたスラッグに一致するカテゴリを取得
 * @param slug カテゴリのスラッグ
 * @returns カテゴリオブジェクト、存在しない場合はnull
 */
export async function getCorporateNewsCategoryBySlug(
  slug: string
): Promise<CorporateNewsCategory | null> {
  const response = await client.getList({
    endpoint: 'corporate_news_categories',
    queries: { filters: `slug[equals]${slug}`, limit: 1 },
  })
  const validated = corporateNewsCategoryListSchema.parse(response)
  return validated.contents[0] || null
}

/**
 * すべてのカテゴリのスラッグを取得
 * @doc 静的生成のためのスラッグ一覧取得
 * @returns スラッグの配列
 */
export async function getAllCorporateNewsCategorySlugs(): Promise<string[]> {
  const allCategories = await getAllContents<{ id: string; slug: string }>(
    'corporate_news_categories',
    {
      fields: 'id,slug',
    }
  )
  return allCategories.map((c) => c.slug)
}