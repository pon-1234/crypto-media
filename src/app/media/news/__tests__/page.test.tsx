/**
 * @vitest-environment jsdom
 */
import { describe, it, expect, vi } from 'vitest'
import MediaNewsPage from '../page'
import { getMediaArticlesByType, getCategories, getTags } from '@/lib/microcms'
import type { MediaArticle, Category, Tag } from '@/lib/schema'

vi.mock('@/lib/microcms', () => ({
  getMediaArticlesByType: vi.fn(),
  getCategories: vi.fn(),
  getTags: vi.fn(),
}))

vi.mock('@/components/media/ArticleCard', () => ({
  ArticleCard: vi.fn(({ article }) => (
    <div data-testid={`article-${article.id}`}>{article.title}</div>
  )),
}))

vi.mock('@/components/ui/Pagination', () => ({
  Pagination: vi.fn(({ currentPage, totalPages }) => (
    <div data-testid="pagination">
      Page {currentPage} of {totalPages}
    </div>
  )),
}))

vi.mock('@/components/media/ArticleFilters', () => ({
  ArticleFilters: vi.fn(({ selectedCategory, selectedTag }) => (
    <div data-testid="filters">
      Category: {selectedCategory || 'none'}, Tag: {selectedTag || 'none'}
    </div>
  )),
}))

describe('MediaNewsPage', () => {
  const mockArticles: MediaArticle[] = [
    {
      id: '1',
      title: 'メディアニュース1',
      slug: 'news-1',
      type: 'media_news',
      membershipLevel: 'public',
      content: 'ニュース内容1',
      heroImage: {
        url: 'https://example.com/image1.jpg',
        width: 1200,
        height: 630,
      },
      createdAt: '2024-01-01T00:00:00Z',
      updatedAt: '2024-01-01T00:00:00Z',
      publishedAt: '2024-01-01T00:00:00Z',
      revisedAt: '2024-01-01T00:00:00Z',
    },
    {
      id: '2',
      title: 'メディアニュース2',
      slug: 'news-2',
      type: 'media_news',
      membershipLevel: 'public',
      content: 'ニュース内容2',
      heroImage: {
        url: 'https://example.com/image2.jpg',
        width: 1200,
        height: 630,
      },
      createdAt: '2024-01-02T00:00:00Z',
      updatedAt: '2024-01-02T00:00:00Z',
      publishedAt: '2024-01-02T00:00:00Z',
      revisedAt: '2024-01-02T00:00:00Z',
    },
  ]

  const mockCategories: Category[] = [
    {
      id: 'cat1',
      name: 'Announcement',
      slug: 'announcement',
      createdAt: '2024-01-01T00:00:00Z',
      updatedAt: '2024-01-01T00:00:00Z',
    },
  ]

  const mockTags: Tag[] = [
    {
      id: 'tag1',
      name: 'Update',
      slug: 'update',
      createdAt: '2024-01-01T00:00:00Z',
      updatedAt: '2024-01-01T00:00:00Z',
    },
  ]

  it('メディアニュース一覧を正しく表示する', async () => {
    vi.mocked(getMediaArticlesByType).mockResolvedValue({
      contents: mockArticles,
      totalCount: 2,
      offset: 0,
      limit: 12,
    })
    vi.mocked(getCategories).mockResolvedValue({
      contents: mockCategories,
      totalCount: 1,
      offset: 0,
      limit: 100,
    })
    vi.mocked(getTags).mockResolvedValue({
      contents: mockTags,
      totalCount: 1,
      offset: 0,
      limit: 100,
    })

    const page = await MediaNewsPage({ searchParams: {} })

    expect(getMediaArticlesByType).toHaveBeenCalledWith('media_news', {
      limit: 12,
      offset: 0,
      orders: '-publishedAt',
      filters: undefined,
    })
    expect(page).toMatchSnapshot()
  })

  it('ページネーションが正しく動作する', async () => {
    vi.mocked(getMediaArticlesByType).mockResolvedValue({
      contents: mockArticles,
      totalCount: 30,
      offset: 12,
      limit: 12,
    })
    vi.mocked(getCategories).mockResolvedValue({
      contents: mockCategories,
      totalCount: 1,
      offset: 0,
      limit: 100,
    })
    vi.mocked(getTags).mockResolvedValue({
      contents: mockTags,
      totalCount: 1,
      offset: 0,
      limit: 100,
    })

    const page = await MediaNewsPage({ searchParams: { page: '2' } })

    expect(getMediaArticlesByType).toHaveBeenCalledWith('media_news', {
      limit: 12,
      offset: 12,
      orders: '-publishedAt',
      filters: undefined,
    })
    expect(page).toMatchSnapshot()
  })

  it('カテゴリとタグの複合フィルタが適用される', async () => {
    vi.mocked(getCategories)
      .mockResolvedValueOnce({
        contents: [mockCategories[0]],
        totalCount: 1,
        offset: 0,
        limit: 1,
      })
      .mockResolvedValueOnce({
        contents: mockCategories,
        totalCount: 1,
        offset: 0,
        limit: 100,
      })

    vi.mocked(getTags)
      .mockResolvedValueOnce({
        contents: [mockTags[0]],
        totalCount: 1,
        offset: 0,
        limit: 1,
      })
      .mockResolvedValueOnce({
        contents: mockTags,
        totalCount: 1,
        offset: 0,
        limit: 100,
      })

    vi.mocked(getMediaArticlesByType).mockResolvedValue({
      contents: mockArticles,
      totalCount: 2,
      offset: 0,
      limit: 12,
    })

    await MediaNewsPage({
      searchParams: { category: 'announcement', tag: 'update' },
    })

    expect(getCategories).toHaveBeenCalledWith({
      filters: 'slug[equals]announcement',
    })
    expect(getTags).toHaveBeenCalledWith({ filters: 'slug[equals]update' })
    expect(getMediaArticlesByType).toHaveBeenCalledWith('media_news', {
      limit: 12,
      offset: 0,
      orders: '-publishedAt',
      filters:
        'type[equals]media_news[and]category[equals]cat1[and]tags[contains]tag1',
    })
  })

  it('記事が0件の場合、メッセージを表示する', async () => {
    vi.mocked(getMediaArticlesByType).mockResolvedValue({
      contents: [],
      totalCount: 0,
      offset: 0,
      limit: 12,
    })
    vi.mocked(getCategories).mockResolvedValue({
      contents: mockCategories,
      totalCount: 1,
      offset: 0,
      limit: 100,
    })
    vi.mocked(getTags).mockResolvedValue({
      contents: mockTags,
      totalCount: 1,
      offset: 0,
      limit: 100,
    })

    const page = await MediaNewsPage({ searchParams: {} })

    expect(page).toMatchSnapshot()
  })

  it('エラーが発生した場合、エラーをスローする', async () => {
    const error = new Error('API Error')
    vi.mocked(getMediaArticlesByType).mockRejectedValue(error)

    await expect(MediaNewsPage({ searchParams: {} })).rejects.toThrow(
      'API Error'
    )
  })
})
