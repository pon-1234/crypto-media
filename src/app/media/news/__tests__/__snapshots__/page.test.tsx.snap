// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MediaNewsPage > ページネーションが正しく動作する 1`] = `
<div
  className="min-h-screen bg-gray-50"
>
  <div
    className="mx-auto max-w-7xl px-4 py-8 sm:px-6 sm:py-12 lg:px-8"
  >
    <div
      className="mb-8 sm:mb-12"
    >
      <h1
        className="mb-4 text-2xl font-bold text-gray-900 sm:text-3xl md:text-4xl"
      >
        ニュース・お知らせ
      </h1>
      <p
        className="max-w-3xl text-base text-gray-600 sm:text-lg"
      >
        暗号資産・ブロックチェーン業界の最新ニュースとお知らせをお届けします。
      </p>
    </div>
    <div
      className="mb-8 lg:mb-12"
    >
      <spy
        categories={
          [
            {
              "createdAt": "2024-01-01T00:00:00Z",
              "id": "cat1",
              "name": "Announcement",
              "slug": "announcement",
              "updatedAt": "2024-01-01T00:00:00Z",
            },
          ]
        }
        tags={
          [
            {
              "createdAt": "2024-01-01T00:00:00Z",
              "id": "tag1",
              "name": "Update",
              "slug": "update",
              "updatedAt": "2024-01-01T00:00:00Z",
            },
          ]
        }
      />
    </div>
    <React.Fragment>
      <div
        className="mb-8 grid grid-cols-1 gap-6 sm:gap-8 md:grid-cols-2 lg:grid-cols-3"
      >
        <spy
          article={
            {
              "content": "ニュース内容1",
              "createdAt": "2024-01-01T00:00:00Z",
              "heroImage": {
                "height": 630,
                "url": "https://example.com/image1.jpg",
                "width": 1200,
              },
              "id": "1",
              "membershipLevel": "public",
              "publishedAt": "2024-01-01T00:00:00Z",
              "revisedAt": "2024-01-01T00:00:00Z",
              "slug": "news-1",
              "title": "メディアニュース1",
              "type": "media_news",
              "updatedAt": "2024-01-01T00:00:00Z",
            }
          }
        />
        <spy
          article={
            {
              "content": "ニュース内容2",
              "createdAt": "2024-01-02T00:00:00Z",
              "heroImage": {
                "height": 630,
                "url": "https://example.com/image2.jpg",
                "width": 1200,
              },
              "id": "2",
              "membershipLevel": "public",
              "publishedAt": "2024-01-02T00:00:00Z",
              "revisedAt": "2024-01-02T00:00:00Z",
              "slug": "news-2",
              "title": "メディアニュース2",
              "type": "media_news",
              "updatedAt": "2024-01-02T00:00:00Z",
            }
          }
        />
      </div>
      <div
        className="mt-8"
      >
        <div
          className="mb-4 text-center"
        >
          <p
            className="text-sm text-gray-600 sm:text-base"
          >
            全
            30
            件中
            13
            -
            14
            件を表示
          </p>
        </div>
        <spy
          className="mt-8"
          currentPage={2}
          totalPages={3}
        />
      </div>
    </React.Fragment>
  </div>
</div>
`;

exports[`MediaNewsPage > メディアニュース一覧を正しく表示する 1`] = `
<div
  className="min-h-screen bg-gray-50"
>
  <div
    className="mx-auto max-w-7xl px-4 py-8 sm:px-6 sm:py-12 lg:px-8"
  >
    <div
      className="mb-8 sm:mb-12"
    >
      <h1
        className="mb-4 text-2xl font-bold text-gray-900 sm:text-3xl md:text-4xl"
      >
        ニュース・お知らせ
      </h1>
      <p
        className="max-w-3xl text-base text-gray-600 sm:text-lg"
      >
        暗号資産・ブロックチェーン業界の最新ニュースとお知らせをお届けします。
      </p>
    </div>
    <div
      className="mb-8 lg:mb-12"
    >
      <spy
        categories={
          [
            {
              "createdAt": "2024-01-01T00:00:00Z",
              "id": "cat1",
              "name": "Announcement",
              "slug": "announcement",
              "updatedAt": "2024-01-01T00:00:00Z",
            },
          ]
        }
        tags={
          [
            {
              "createdAt": "2024-01-01T00:00:00Z",
              "id": "tag1",
              "name": "Update",
              "slug": "update",
              "updatedAt": "2024-01-01T00:00:00Z",
            },
          ]
        }
      />
    </div>
    <React.Fragment>
      <div
        className="mb-8 grid grid-cols-1 gap-6 sm:gap-8 md:grid-cols-2 lg:grid-cols-3"
      >
        <spy
          article={
            {
              "content": "ニュース内容1",
              "createdAt": "2024-01-01T00:00:00Z",
              "heroImage": {
                "height": 630,
                "url": "https://example.com/image1.jpg",
                "width": 1200,
              },
              "id": "1",
              "membershipLevel": "public",
              "publishedAt": "2024-01-01T00:00:00Z",
              "revisedAt": "2024-01-01T00:00:00Z",
              "slug": "news-1",
              "title": "メディアニュース1",
              "type": "media_news",
              "updatedAt": "2024-01-01T00:00:00Z",
            }
          }
        />
        <spy
          article={
            {
              "content": "ニュース内容2",
              "createdAt": "2024-01-02T00:00:00Z",
              "heroImage": {
                "height": 630,
                "url": "https://example.com/image2.jpg",
                "width": 1200,
              },
              "id": "2",
              "membershipLevel": "public",
              "publishedAt": "2024-01-02T00:00:00Z",
              "revisedAt": "2024-01-02T00:00:00Z",
              "slug": "news-2",
              "title": "メディアニュース2",
              "type": "media_news",
              "updatedAt": "2024-01-02T00:00:00Z",
            }
          }
        />
      </div>
    </React.Fragment>
  </div>
</div>
`;

exports[`MediaNewsPage > 記事が0件の場合、メッセージを表示する 1`] = `
<div
  className="min-h-screen bg-gray-50"
>
  <div
    className="mx-auto max-w-7xl px-4 py-8 sm:px-6 sm:py-12 lg:px-8"
  >
    <div
      className="mb-8 sm:mb-12"
    >
      <h1
        className="mb-4 text-2xl font-bold text-gray-900 sm:text-3xl md:text-4xl"
      >
        ニュース・お知らせ
      </h1>
      <p
        className="max-w-3xl text-base text-gray-600 sm:text-lg"
      >
        暗号資産・ブロックチェーン業界の最新ニュースとお知らせをお届けします。
      </p>
    </div>
    <div
      className="mb-8 lg:mb-12"
    >
      <spy
        categories={
          [
            {
              "createdAt": "2024-01-01T00:00:00Z",
              "id": "cat1",
              "name": "Announcement",
              "slug": "announcement",
              "updatedAt": "2024-01-01T00:00:00Z",
            },
          ]
        }
        tags={
          [
            {
              "createdAt": "2024-01-01T00:00:00Z",
              "id": "tag1",
              "name": "Update",
              "slug": "update",
              "updatedAt": "2024-01-01T00:00:00Z",
            },
          ]
        }
      />
    </div>
    <div
      className="py-16 text-center sm:py-24"
    >
      <p
        className="text-gray-600"
      >
        ニュースがありません。
      </p>
    </div>
  </div>
</div>
`;
